packages/stock-location/src/services/stock-location.ts

```diff
@@ -0,0 +1,266 @@
+   */
+  async listAndCount(
+    selector: FilterableStockLocationProps = {},
+    config: FindConfig<StockLocation> = { relations: [], skip: 0, take: 10 }
+  ): Promise<[StockLocation[], number]> {
+    const manager = this.getManager()
+    const locationRepo = manager.getRepository(StockLocation)
+
+    const query = buildQuery(selector, config)
+    return await locationRepo.findAndCount(query)
+  }
+
+  /**
+   * Retrieves a stock location by its ID.
+   * @param {string} stockLocationId - The ID of the stock location.
+   * @param {FindConfig} [config={}] - Additional configuration for the query.
+   * @return {Promise<StockLocation>} The stock location.
+   * @throws {MedusaError} If the stock location ID is not defined.
+   * @throws {MedusaError} If the stock location with the given ID was not found.
+   */
+  async retrieve(
+    stockLocationId: string,
+    config: FindConfig<StockLocation> = {}
+  ): Promise<StockLocation> {
+    if (!isDefined(stockLocationId)) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `"stockLocationId" must be defined`
+      )
+    }
+
+    const manager = this.getManager()
+    const locationRepo = manager.getRepository(StockLocation)
+
+    const query = buildQuery({ id: stockLocationId }, config)
+    const loc = await locationRepo.findOne(query)
+
+    if (!loc) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `StockLocation with id ${stockLocationId} was not found`
+      )
+    }
+
+    return loc
+  }
+
+  /**
+   * Creates a new stock location.
+   * @param {CreateStockLocationInput} data - The input data for creating a stock location.
+   * @returns {Promise<StockLocation>} - The created stock location.
+   */
+  async create(data: CreateStockLocationInput): Promise<StockLocation> {
+    const defaultManager = this.getManager()
+    return await defaultManager.transaction(async (manager) => {
+      const locationRepo = manager.getRepository(StockLocation)
+
+      const loc = locationRepo.create({
+        name: data.name,
+      })
+
+      if (isDefined(data.address) || isDefined(data.address_id)) {
+        if (typeof data.address === "string" || data.address_id) {
+          const addrId = (data.address ?? data.address_id) as string
+          const address = await this.retrieve(addrId, {
+            select: ["id"],
+          })
+          loc.address_id = address.id
+        } else {
+          const locAddressRepo = manager.getRepository(StockLocationAddress)
+          const locAddress = locAddressRepo.create(data.address!)
+          const addressResult = await locAddressRepo.save(locAddress)
+          loc.address_id = addressResult.id
+        }
+      }
+
+      const { metadata } = data
+      if (metadata) {
+        loc.metadata = setMetadata(loc, metadata)
+      }
+
+      const result = await locationRepo.save(loc)
+
+      await this.eventBusService_.emit(StockLocationService.Events.CREATED, {
```

Comment:
This will need to be properly handled by the Event Outbox pattern that @olivermrbl is planning.

Reply:
Can we add a todo? My concern is that since this feature is not guarded, anybody can install the module and end up with discrepancies. Do we have a warning somewhere ? I might have missed it